Add a CRaC packaging type to create checkpointed docker images

timyates opened the pull request:
This allows you to do

    mvn package -Dpackaging=docker-crac

I am still working on publish

alvarosanchez commented on src/main/java/io/micronaut/build/DockerCracMojo.java:
```diff
@@ -0,0 +1,206 @@
+        return name;
+    }
+
+    private String getReadinessCommand() {
+        String property = mavenProject.getProperties().getProperty(DockerCracMojo.CRAC_READINESS_PROPERTY);
+        return property == null ? DockerCracMojo.DEFAULT_READINESS_COMMAND : property;
+    }
```
You should declare a field and annotate it with:

```java
@Parameter(property = DockerCracMojo.CRAC_READINESS_PROPERTY, defaultValue = DockerCracMojo.DEFAULT_READINESS_COMMAND)
```

alvarosanchez commented on src/main/java/io/micronaut/build/DockerCracMojo.java:
```diff
@@ -0,0 +1,206 @@
+    private void copyScripts(String... scriptNames) throws IOException {
+        List<String> imageClasspathScopes = Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_RUNTIME);
+        String readinessCommand = getReadinessCommand();
+        mavenProject.setArtifactFilter(artifact -> imageClasspathScopes.contains(artifact.getScope()));
```
Why is this needed?

timyates replied:
I copied it from somewhere...  My lack of Maven skills is showing 😉 Gone now

alvarosanchez commented on src/main/java/io/micronaut/build/DockerCracMojo.java:
```diff
@@ -0,0 +1,206 @@
+        processScripts(target, s -> s.replace("@READINESS@", readinessCommand).replace("@MAINCLASS@", mainClass), scriptNames);
+    }
+
+    private void processScripts(File target, UnaryOperator<String> replacement, String... scriptNames) throws IOException {
+        for (String script : scriptNames) {
+            File localOverride = new File(mavenProject.getBasedir(), script);
+            try (InputStream resourceAsStream = localOverride.exists() ?
+                    Files.newInputStream(localOverride.toPath()) :
+                    DockerCracMojo.class.getResourceAsStream("/cracScripts/" + script)) {
+                if (resourceAsStream == null) {
+                    throw new IOException("Could not find script " + script);
+                }
+                List<String> tokenized = replaceTokensInTextStream(resourceAsStream, replacement);
+                Path outputPath = target.toPath().resolve(script);
+                writeStringsToFile(tokenized, outputPath);
+                Files.setPosixFilePermissions(outputPath, POSIX_FILE_PERMISSIONS);
+            }
+        }
+    }
+
+    private static List<String> replaceTokensInTextStream(InputStream resourceAsStream, UnaryOperator<String> replacement) throws IOException {
+        return IOUtils.readLines(resourceAsStream, StandardCharsets.UTF_8)
+                .stream()
+                .map(replacement)
+                .collect(Collectors.toList());
+    }
+
+    private static void writeStringsToFile(List<String> tokenized, Path outputPath) throws IOException {
+        try (BufferedWriter out = Files.newBufferedWriter(outputPath, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+            IOUtils.writeLines(tokenized, System.getProperty("line.separator"), out);
+        }
+    }
```
There is a better, Maven-ish way to do this, which is using the Maven Resources Plugin API: https://maven.apache.org/shared/maven-filtering/usage.html

You should be able to inject `MavenResourcesFiltering` in the constructor.

timyates replied:
Ok, so I tried this but failed hard

How do you use that to copy a classpath resource and replace tokens?  I can get it to copy a file, but not extract the files from the classpath of the mojo

Any clues would be handy...

alvarosanchez replied:
Perhaps you can use `MavenReaderFilter` instead. Something like this:

```java
InputStreamReader reader = new InputStreamReader(anInputStream); //classpath resource
MavenReaderFilter readerFilter; //inject it
MavenReaderFilterRequest req = new MavenReaderFilterRequest();
Properties additionalProperties = new Properties();
additionalProperties.setProperty("READINESS", "..." );
req.setFrom(reader);
req.setFiltering(true);
req.setAdditionalProperties(additionalProperties);
String result = IOUtils.toString(readerFilter.filter(req));
IOUtils.write(result, out); //write to file...
```

I got this from https://github.com/apache/maven-filtering/blob/master/src/test/java/org/apache/maven/shared/filtering/DefaultMavenReaderFilterTest.java

Take a look at the other tests in case you see something better.

alvarosanchez commented on src/main/java/io/micronaut/build/services/DockerService.java:
```diff
@@ -104,6 +114,36 @@ public void onNext(BuildResponseItem item) {
+        try (CreateContainerCmd create = dockerClient.createContainerCmd(imageId)) {
+            HostConfig hostConfig = create.getHostConfig();
+            if (hostConfig == null) {
+                LOG.error("HostConfig is null");
```
Can this message be improved? Perhaps a hint to the user about what happened or what to do?

timyates replied:
This should never be null (though I guess it _can_ be), and if it is, it's kinda catastrophic...

I've switched to throwing an exception here, recommending people check their docker installation 🤔

alvarosanchez commented on src/main/java/io/micronaut/build/services/DockerService.java:
```diff
@@ -104,6 +114,36 @@ public void onNext(BuildResponseItem item) {
+                LOG.info("Container started: {} {}", createResponse.getId(), start.getContainerId());
+                try (WaitContainerCmd wait = dockerClient.waitContainerCmd(createResponse.getId())) {
+                    WaitContainerResultCallback waitResult = wait.start();
+                    Integer exitcode = waitResult.awaitStatusCode(1, TimeUnit.MINUTES);
```
Would it make sense for this timeout to be configurable? Or is it not worth it?

alvarosanchez commented:
@timyates great job here! 🚀 

Apart from the review comments, I miss the following:
- Support for `mvn mn:dockerfile -Dpackaging=docker-crac`, and its corresponding test.
- Does this work with AOT? If so, can you please add a test as well?

timyates commented on src/it/package-docker-crac/pom.xml:
```diff
@@ -0,0 +1,168 @@
+    <dependency>
+      <groupId>io.micronaut.crac</groupId>
+      <artifactId>micronaut-crac</artifactId>
+      <version>1.0.1</version>
```
It is, but v1.0.0 (in the BOM) doesn't work

We had to release 1.0.0 for 3.7.0, but we required a change in the netty server that is used in 1.0.1

Once 3.7.1 is out, and this changes to use it, I will remove this version from this pom

alvarosanchez replied:
You can set `it.micronaut.version` to 3.7.1-SNAPSHOT instead no?

timyates replied:
I had to bump `it.micronaut.version` to 3.7.0-SNAPSHOT (was 3.6.0-SNAPSHOT)

timyates replied:
Yes 😊 

And then I could get rid of the 1.0.1 above 🎉

timyates commented on src/it/package-docker-crac/src/main/resources/logback.xml:
```diff
@@ -0,0 +1,15 @@
+        </encoder>
+    </appender>
+
+    <root level="debug">
```
👍

timyates commented:
@alvarosanchez Thanks! 😊 

Ok, I think I addressed everything above apart from using `MavenResourcesFiltering`.  I tried using it, but the code ended up being much more complex, AND it didn't work...

I must be doing something wrong...

Anyway, TL;DR:

- Up to micronaut 3.7.1-SNAPSHOT and got rid of CRaC [versioning](https://github.com/micronaut-projects/micronaut-maven-plugin/pull/497/commits/35eb829d2be36b884817e69464d6d2ad6c77eb4d)
- Made checkpoint [timeout configurable](https://github.com/micronaut-projects/micronaut-maven-plugin/pull/497/commits/ea49afec566e83f3d89b60b2d9a679448c03c036)
- Added [AOT test](https://github.com/micronaut-projects/micronaut-maven-plugin/pull/497/commits/a4610c2d5904b4228f3c1916d97bcf25b99bacef)
- Enabled `mn:dockerfile` -- this [generates 2 dockerfiles](https://github.com/micronaut-projects/micronaut-maven-plugin/pull/497/commits/5fbe9f120b57df69430616f7d490723103eef010)
- Added a link to CRaC [in the docs](https://github.com/micronaut-projects/micronaut-maven-plugin/pull/497/commits/5a58c160a422a683aff6f076aa3dc45c9e763d91)
- Defined props via a [Maven Parameter](https://github.com/micronaut-projects/micronaut-maven-plugin/pull/497/commits/99a72b3a588c92afcea4d00dee941ef0b179ad5c)

timyates commented:
@alvarosanchez 

That worked...  ([2 commits](https://github.com/micronaut-projects/micronaut-maven-plugin/pull/497/files/050a1b01eded5c0d7a6c591dc596a2703a281365..3df7fc1b82cf08e7faddffe284e42754e397c85c))

alvarosanchez commented:
@timyates thank you! I'll merge it, need to accommodate master for 3.5 first.

alvarosanchez commented:
I have reverted #495 because of https://github.com/graalvm/native-build-tools/issues/327
